### PR TITLE
Add Farach algorithm for suffix tree and suffix array

### PR DESCRIPTION
**Pull request by bart1e**
Implementacja algorytmu Faracha wraz z opisem. Prosiłbym o ewentualne sugestie co poprawić. Jeśli testy dla drzew / tablic sufiksowych dostępne już w repozytorium wystarczą, można usunąć plik z dodatkowymi, małymi testami, które stworzyłem.
Bartosz Wodziński

### CODE REVIEW

File: string_indexing/farach_suffix_tree.py [hunk 1]
```diff
@@ -0,0 +1,427 @@
...
+        suffix = text[A[i] - 1:]
+        currentNode = lastNode
+        while currentNode.depth > LCP[i - 1]:
+            currentNode = currentNode.parent
+        if currentNode.depth == LCP[i - 1]:
+            if currentNode.depth == len(suffix):
+                currentNode.index = len(text) - len(suffix) + 1
+                # dummy object creation; won't have any effect
+                # done to suppress warning 'newNode might be referenced before assignment' at the end of the for loop
+                newNode = trie.TrieNode('')
```
[the remark this answers is not in the document]
Comment (bart1e) [reply string_indexing/farach_suffix_tree.py]: Thank you for the review. I made the changes you have requested. At this place, it is even better to do newNode = currentNode, I think. I will run make check, submit changes and will ask for another review.

**Comment by bart1e** [comment]
Thank you for another review. I will fix remaining issues. As for unused arguments, they are in methods ensuring compatibility with already existing tests, so they have to (I think) take an extra parameter n, which is not used. Is there any elegant way to fix this problem?

**Comment by bart1e** [comment]
I have done my best to correct remaining issues. The only ones left reported by pylint are these related to 'too many branches/statements/local variables' and those described above.

**Comment by krzysztof-turowski** [comment]
> As for unused arguments, they are in methods ensuring compatibility with already existing tests, so they have to (I think) take an extra parameter n, which is not used. Is there any elegant way to fix this problem?
It's done using underscore. Anyway, I fixed it (see L192 in farach_suffix_tree.py)

**Pull request merged**